In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [17]:
df_2021 = pd.read_csv(r'C:\Users\Praveen Shahani\Desktop\Datasets\2021_OPM_FEVS_PRDF.csv')

In [3]:
## Checking if there is any duplicate values
df.drop_duplicates()

RandomID agency   Q1   Q2   Q3   Q4 Q5 Q6 Q7 Q8  ...  DRNO DHISP  \
0       112970976817     XX  5.0  5.0  5.0  5.0  5  4  4  4  ...     A     B   
1       194868625278     XX  3.0  2.0  4.0  3.0  4  2  4  2  ...   NaN     A   
2       152966380283     XX  5.0  5.0  4.0  4.0  3  4  5  4  ...     B     B   
3       193041162980     XX  5.0  5.0  5.0  5.0  5  5  5  5  ...     B     B   
4       146655962451     XX  4.0  5.0  5.0  4.0  4  3  5  4  ...     B     B   
...              ...    ...  ...  ...  ...  ... .. .. .. ..  ...   ...   ...   
292515  154057939422     ST  3.0  4.0  4.0  3.0  4  4  2  4  ...     B     B   
292516  151758964104     ST  3.0  4.0  2.0  4.0  2  2  3  3  ...     B     B   
292517  143492802997     ST  3.0  3.0  NaN  4.0  4  4  4  3  ...   NaN   NaN   
292518  110267537558     ST  2.0  4.0  5.0  5.0  2  1  5  4  ...     B     B   
292519  153195177928     ST  4.0  4.0  5.0  5.0  4  3  5  4  ...     B     B   

       DDIS DAGEGRP DSUPER DFEDTEN DSEX DMIL DLEAVING    POSTWT  
0         B       B      B       A    A    A        A  2.209652  
1         B       B      B       B    A    A        C  2.209652  
2         B       B      A       B    A    B        C  1.858874  
3         A       B      B       B    A    A        A  1.228573  
4         B       B      B       B    A    A        A  1.735842  
...     ...     ...    ...     ...  ...  ...      ...       ...  
292515    B       A      A       A    B    B        A  3.004992  
292516    B       A      A       A    B    B        D  4.427855  
292517  NaN     NaN    NaN       A    A  NaN        A  4.202227  
292518    B       A      A       A    B    B        C  3.523113  
292519    A       B      A       A    A    B        A  3.400784  

[292520 rows x 79 columns]

In [4]:
## Checking how many data values are missing from the dataframe
df.isnull().sum().sum()

754297

In [5]:
## A check to understand the data types of every column
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 292520 entries, 0 to 292519
Data columns (total 79 columns):
 #   Column    Non-Null Count   Dtype  
---  ------    --------------   -----  
 0   RandomID  292520 non-null  int64  
 1   agency    292520 non-null  object 
 2   Q1        291074 non-null  float64
 3   Q2        288624 non-null  float64
 4   Q3        290112 non-null  float64
 5   Q4        291324 non-null  float64
 6   Q5        290770 non-null  object 
 7   Q6        286449 non-null  object 
 8   Q7        290240 non-null  object 
 9   Q8        291451 non-null  object 
 10  Q9        291910 non-null  float64
 11  Q10       292038 non-null  object 
 12  Q11       292018 non-null  object 
 13  Q12       291755 non-null  object 
 14  Q13       291920 non-null  object 
 15  Q14       291306 non-null  object 
 16  Q15       288257 non-null  object 
 17  Q16       290367 non-null  object 
 18  Q17       290268 non-null  object 
 19  Q18       291274 non-null  object 
 20  Q19 

In [6]:
df.Q11.value_counts()
#df.Q10.value_counts()
df.Q20.value_counts()

4    116393
5     76580
3     43704
2     29844
1     17361
X      6044
Name: Q20, dtype: int64

In [7]:
objects= [column for column, is_type in (df.dtypes=="object").items() if is_type]
print(objects)
## The column 'agency' holds no significance here so it is to be dropped.
objects.remove('agency')

['agency', 'Q5', 'Q6', 'Q7', 'Q8', 'Q10', 'Q11', 'Q12', 'Q13', 'Q14', 'Q15', 'Q16', 'Q17', 'Q18', 'Q19', 'Q20', 'Q21', 'Q22', 'Q24', 'Q25', 'Q26', 'Q27', 'Q32', 'Q33', 'Q34', 'Q35', 'Q36', 'Q37', 'Q38', 'Q48', 'Q49', 'Q50', 'Q51', 'Q52', 'Q53', 'Q55', 'Q56', 'Q57', 'DRNO', 'DHISP', 'DDIS', 'DAGEGRP', 'DSUPER', 'DFEDTEN', 'DSEX', 'DMIL', 'DLEAVING']


In [8]:
import numpy as np
for i in objects:
    df[i] = df[i].replace(['X'], np.nan)

In [9]:
df.isnull().sum().sum()

1166370

Further it is required to downcast the object values to integar values to make it easier for plotting graphs

In [10]:
objects[:-9]
for i in objects[:-9]:
    df[i] = df[i].astype(float)

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 292520 entries, 0 to 292519
Data columns (total 79 columns):
 #   Column    Non-Null Count   Dtype  
---  ------    --------------   -----  
 0   RandomID  292520 non-null  int64  
 1   agency    292520 non-null  object 
 2   Q1        291074 non-null  float64
 3   Q2        288624 non-null  float64
 4   Q3        290112 non-null  float64
 5   Q4        291324 non-null  float64
 6   Q5        290329 non-null  float64
 7   Q6        285707 non-null  float64
 8   Q7        289507 non-null  float64
 9   Q8        282470 non-null  float64
 10  Q9        291910 non-null  float64
 11  Q10       245119 non-null  float64
 12  Q11       219694 non-null  float64
 13  Q12       265187 non-null  float64
 14  Q13       289170 non-null  float64
 15  Q14       280411 non-null  float64
 16  Q15       281489 non-null  float64
 17  Q16       283480 non-null  float64
 18  Q17       283768 non-null  float64
 19  Q18       286829 non-null  float64
 20  Q19 

It comes out that the missing values are much more than what we have initially found out

### Handling missing values

There are two most common ways to deal with missing values:

1. Deleting the entire row of the missing values

2. Replacing the missing value from the relevant measure of central tendency (mean, median, mode)

In [12]:
df_c = df.copy()
df_c

RandomID agency   Q1   Q2   Q3   Q4   Q5   Q6   Q7   Q8  ...  \
0       112970976817     XX  5.0  5.0  5.0  5.0  5.0  4.0  4.0  4.0  ...   
1       194868625278     XX  3.0  2.0  4.0  3.0  4.0  2.0  4.0  2.0  ...   
2       152966380283     XX  5.0  5.0  4.0  4.0  3.0  4.0  5.0  4.0  ...   
3       193041162980     XX  5.0  5.0  5.0  5.0  5.0  5.0  5.0  5.0  ...   
4       146655962451     XX  4.0  5.0  5.0  4.0  4.0  3.0  5.0  4.0  ...   
...              ...    ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
292515  154057939422     ST  3.0  4.0  4.0  3.0  4.0  4.0  2.0  4.0  ...   
292516  151758964104     ST  3.0  4.0  2.0  4.0  2.0  2.0  3.0  3.0  ...   
292517  143492802997     ST  3.0  3.0  NaN  4.0  4.0  4.0  4.0  3.0  ...   
292518  110267537558     ST  2.0  4.0  5.0  5.0  2.0  1.0  5.0  4.0  ...   
292519  153195177928     ST  4.0  4.0  5.0  5.0  4.0  3.0  5.0  4.0  ...   

        DRNO  DHISP  DDIS  DAGEGRP  DSUPER  DFEDTEN  DSEX  DMIL  DLEAVING  \
0          A      B     B        B       B        A     A     A         A   
1        NaN      A     B        B       B        B     A     A         C   
2          B      B     B        B       A        B     A     B         C   
3          B      B     A        B       B        B     A     A         A   
4          B      B     B        B       B        B     A     A         A   
...      ...    ...   ...      ...     ...      ...   ...   ...       ...   
292515     B      B     B        A       A        A     B     B         A   
292516     B      B     B        A       A        A     B     B         D   
292517   NaN    NaN   NaN      NaN     NaN        A     A   NaN         A   
292518     B      B     B        A       A        A     B     B         C   
292519     B      B     A        B       A        A     A     B         A   

          POSTWT  
0       2.209652  
1       2.209652  
2       1.858874  
3       1.228573  
4       1.735842  
...          ...  
292515  3.004992  
292516  4.427855  
292517  4.202227  
292518  3.523113  
292519  3.400784  

[292520 rows x 79 columns]

### Starting with deleting the entire row

In [13]:
df_c = df_c.dropna()
df_c

RandomID agency   Q1   Q2   Q3   Q4   Q5   Q6   Q7   Q8  ...  \
3       193041162980     XX  5.0  5.0  5.0  5.0  5.0  5.0  5.0  5.0  ...   
5       195312241136     XX  4.0  3.0  4.0  4.0  4.0  4.0  4.0  3.0  ...   
6       187815423862     XX  5.0  4.0  4.0  4.0  4.0  4.0  4.0  4.0  ...   
9       138139604049     XX  4.0  4.0  5.0  4.0  4.0  4.0  4.0  4.0  ...   
12      135019931353     XX  4.0  4.0  5.0  5.0  2.0  2.0  4.0  4.0  ...   
...              ...    ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
292503  199379325377     ST  1.0  1.0  1.0  2.0  1.0  1.0  1.0  1.0  ...   
292513  102229919658     ST  5.0  5.0  5.0  4.0  2.0  5.0  5.0  5.0  ...   
292514  117508083776     ST  3.0  4.0  4.0  4.0  3.0  4.0  4.0  3.0  ...   
292516  151758964104     ST  3.0  4.0  2.0  4.0  2.0  2.0  3.0  3.0  ...   
292518  110267537558     ST  2.0  4.0  5.0  5.0  2.0  1.0  5.0  4.0  ...   

        DRNO  DHISP  DDIS  DAGEGRP  DSUPER  DFEDTEN  DSEX  DMIL  DLEAVING  \
3          B      B     A        B       B        B     A     A         A   
5          A      B     B        B       A        B     B     A         A   
6          B      B     B        B       B        B     A     A         A   
9          B      B     A        B       B        A     A     A         A   
12         B      B     B        B       B        B     B     B         A   
...      ...    ...   ...      ...     ...      ...   ...   ...       ...   
292503     C      B     B        A       B        A     A     B         A   
292513     B      A     B        A       A        A     A     A         A   
292514     B      B     B        A       A        A     A     B         A   
292516     B      B     B        A       A        A     B     B         D   
292518     B      B     B        A       A        A     B     B         C   

          POSTWT  
3       1.228573  
5       2.076046  
6       1.858874  
9       1.228573  
12      1.089288  
...          ...  
292503  5.774132  
292513  5.476801  
292514  3.487832  
292516  4.427855  
292518  3.523113  

[99128 rows x 79 columns]

In [14]:
df_c.isnull().sum().sum() ##-> No missing values found in the dataset

0

Having significant change (in the size of the matrix) in the missing values, we may require to check how many survey entries has been altered 

In [15]:
print('The number of survey entries altered:',df.shape[0]-df_c.shape[0]) 
print('The percentage of alterations from the original dataset is: ',round(((df.shape[0]-df_c.shape[0])/(df.shape[0])),2)*100,'%')

The number of survey entries altered: 193392
The percentage of alterations from the original dataset is:  66.0 %


66% of the alteration of data is a significant impact. There is surely a requirenment for imputation which is replacing the missing values by the potential measure of central tendency. 

But the imputed value may not rightly represent the interest of the respondent. Therefore, keeping the data natural as much as possible.